app/src/main/java/org/fossasia/susi/ai/signup/SignUpInteractor.kt

```diff
@@ -0,0 +1,95 @@
+package org.fossasia.susi.ai.signup
+
+import org.fossasia.susi.ai.helper.Constant
+import org.fossasia.susi.ai.helper.CredentialHelper
+import org.fossasia.susi.ai.helper.PrefManager
+import org.fossasia.susi.ai.rest.ClientBuilder
+import org.fossasia.susi.ai.rest.responses.susi.SignUpResponse
+import retrofit2.Call
+import retrofit2.Callback
+import retrofit2.Response
+import java.net.UnknownHostException
+
+/**
+ * Created by mayanktripathi on 05/07/17.
+ */
+class SignUpInteractor: ISignUpInteractor {
+
+    override fun signUp(email: String, password: String, conpass: String, url: String, listener: ISignUpInteractor.OnLoginFinishedListener) {
+        if(email.isEmpty()){
+            listener.emptyEmail()
+            return
+        }
+        if(password.isEmpty()){
+            listener.emptyPassword()
+            return
+        }
+        if(conpass.isEmpty()){
+            listener.emptyConPassword()
+            return
+        }
+        if (!CredentialHelper.isEmailValid(email)) {
+            listener.setErrorEmail()
+            return
+        }
+        if (!CredentialHelper.checkPasswordValid(password)) {
+            listener.passwordInvalid()
+            return
+        }
+        if (password != conpass) {
+            listener.setErrorPass()
+            return
+        }
+        if(listener.isPersonalServer()) {
```

Comment:
Its not about looking clean. The listener is used as a callback to notify presenter about change in data so that the presenter then may further invoke view to update UI. In this flow data must go from Model to Presenter to View and not from view to presenter to model. Use listener to notify about changing UI and not for getting something from View or presenter.